### PR TITLE
Auth - Mapping of AWSMobileClient error to Auth Error

### PR DESCRIPTION
**Pull request by royjit**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+        case .invalidState(let message):
+            return AmplifyAuthError.invalidState(message, AuthPluginErrorConstants.invalidStateError)
+
+        case .invalidConfiguration(let message),
+             .cognitoIdentityPoolNotConfigured(let message),
+             .invalidUserPoolConfiguration(let message):
+            return AmplifyAuthError.configuration(message, AuthPluginErrorConstants.configurationError)
+
+        case .notAuthorized(let message):
+            // Not authorized is thrown from server at different scenarios
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift:125]: This isn't a helpful comment. What is it telling me that i need to know but can't figure out from the clearly-named case?

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        .unableToSignIn(let message), // Called in signout, releaseSignInWait.
+        .userCancelledSignIn(let message),
+        .idTokenNotIssued(let message), // Not used anywhere.
+        .badRequest(let message), // HostedUI returned with bad url or request.
+        .userPoolNotConfigured(let message):
+            return AmplifyAuthError.unknown(message)
+
+        case .securityFailed(let message):
+            // Error occured inside HostedUI
+            // TODO: Handle this error in Auth.signInWithUI
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift:142]: Here & throughout: If this TODO isn't coming in an imminent PR, add a backlog issue & link it

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+            // Called inside hostedUI showSignIn if there is no idToken or accessToken after getsession completes
+            // This error will be handled by Auth.signInWithUI.
+            // TODO: Handle this error in Auth.signInWithUI
+            return AmplifyAuthError.unknown(message)
+
+        case .expiredRefreshToken(let message):
+            // TODO: Should be handled by Auth.fetchSession
+            return AmplifyAuthError.unknown(message)
+
+        // These errors araise from the escape hatch methods.
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift:155]: s/araise/arise/

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+//
+
+import Amplify
+import AWSMobileClient
+
+struct AuthErrorHelper {
+
+    // swiftlint:disable cyclomatic_complexity
+    // swiftlint:disable function_body_length
+    static func map(awsMobileClientError: AWSMobileClientError) -> AmplifyAuthError {
```
Comment (palpatim) [inline AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift:15]: nit: `map` is usually a term of art used for transforming values in a collection. For one-off transformations, maybe something like 'toAmplifyAuthError()` or similar? Or even constructors like `init(awsMobileClientError: AWSMobileClientError)` and `init(_ error: Error)`? I'm not sure how often we provide initializers for our error cases, though, so aim for consistency with the rest of the codebase here.

File: Amplify/Categories/Auth/Error/AmplifyAuthError.swift [hunk 5]
```diff
@@ -13,19 +13,47 @@ public enum AmplifyAuthError {
...
 
     public var underlyingError: Error? {
-        return nil
+        switch self {
+        case .configuration(_, _, let underlyingError),
+             .service(_, _, let underlyingError),
+             .unknown(_, let underlyingError),
+             .validation(_, _, _, let underlyingError),
+             .notAuthorized(_, _, let underlyingError),
+            .invalidState(_, _, let underlyingError):
```
[the remark this answers is not in the document]
Comment (royjit) [reply Amplify/Categories/Auth/Error/AmplifyAuthError.swift:29]: Moved swiftFormat and swiftLint before compilation step.
But the above change was not caught, manually formatted the file

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Constants/AuthPluginErrorConstants.swift [hunk 6]
```diff
@@ -20,8 +24,69 @@ struct AuthPluginErrorConstants {
...
+
+    static let resourceNotFoundError: RecoverySuggestion = "Make sure that the user pool has a requested resource"
+
+    static let softwareTokenNotFoundError: RecoverySuggestion = "Enable the software token MFA for the user"
+
+    static let tooManyFailedError: RecoverySuggestion = "User might have tried too many times with failed input"
+
+    static let tooManyRequestError: RecoverySuggestion = """
+Make sure the requests send are controlled and the errors are properlly handled
+"""
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/AWSAuthPlugin/Support/Constants/AuthPluginErrorConstants.swift:71]: Fixed

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift (hunk 1 above)
Comment (royjit) [reply AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift:125]: Updated the comments. 
This error is a direct mapping from the service error. And cognito throws this error in many different scenarios like username/pwd is wrong, timeout in signIn etc. That is the reason I initially gave a generic comment.

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+             .cognitoIdentityPoolNotConfigured(let message),
+             .invalidUserPoolConfiguration(let message):
+            return AmplifyAuthError.configuration(message, AuthPluginErrorConstants.configurationError)
+
+        case .notAuthorized(let message):
+            // Not authorized is thrown from server at different scenarios
+            return AmplifyAuthError.notAuthorized(message, AuthPluginErrorConstants.notAuthorizedError)
+
+        // Below error should not happen, these will be handled inside the plugin.
+        case .notSignedIn(let message), // Called in getTokens/getPassword when not signedin to CUP
```
[the remark this answers is not in the document]
Comment (royjit) [reply AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift:129]: Format returned the same indentation.

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift (hunk 2 above)
Comment (royjit) [reply AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift:142]: Added tracking number

File: AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift (hunk 3 above)
Comment (royjit) [reply AmplifyPlugins/Auth/AWSAuthPlugin/Support/Utils/AuthErrorHelper.swift:155]: Fixed

**Pull request merged**